File: BasicML/basicml/optim/sgd.py
from .optimizer             import Optimizer
from typing                 import Optional
from basicml.regularization import Regularizer
from basicml.tensor         import Tensor

class SGD(Optimizer):
    def __init__(self, 
                 parameters : list[Tensor], 
                 lr         : float, 
                 regularizer: Optional[Regularizer] = None):
        super().__init__(parameters, lr, regularizer)

    def step(self):
        for param in self.parameters:
            if param.requires_grad and param.grad is not None:
                grad = param.grad
                if self.regularizer is not None:
                    grad = grad + self.regularizer.grad(param)
                param.data -= self.lr * grad

    def zero_grad(self):
        for param in self.parameters:
            param.zero_grad()

File: AdvanceML/src/optim/sgd.cpp
#include "advanceml/optim/sgd.hpp"

namespace advanceml {

SGD::SGD(std::vector<Tensor> parameters, float learning_rate)
    : parameters_(std::move(parameters)), learning_rate_(learning_rate) {}

void SGD::step() {
    for (Tensor& param : parameters_) {
        if (!param.has_grad()) {
            continue;
        }
        Tensor grad = param.grad();
        FloatBuffer& values = param.mutable_data();
        for (size_t i = 0; i < values.size(); ++i) {
            values[i] -= learning_rate_ * grad.data()[i];
        }
    }
}

void SGD::zero_grad() {
    for (Tensor& param : parameters_) {
        param.zero_grad();
    }
}

}  // namespace advanceml
